tests/kernel/smp/src/main.c

```diff
@@ -224,6 +230,17 @@ static void thread_entry_fn(void *p1, void *p2, void *p3)
 	}
 }
 
+static void thread_entry_inf_fn(void *p1, void *p2, void *p3)
```

Comment:
So... I do not disagree that this is an improvement to the design.  But... it's test code.  And the earlier code wasn't actually wrong, it was just clumsy.  But here's the important part: it was clumsy becuase it was making assumptions that the five 50ms k_busy_wait() calls in the function would not interfere with the smaller delay in test_coop_resched_thread().  And sure: that k_busy_wait() isn't supposed to interfere or be slow or wrong.  And it isn't.  So it's needless.  **In app code**.  But in the test, what's happened is that you've actually removed coverage of k_busy_wait().

Basically, my feeling is that "bad test code is only bad if it's invalid".  Merely "testing the wrong thing" does more good than harm.  There have been a lot of edge cases found via weird test code, historically.

tl;dr: Can you remove this patch?

Reply:
First of all, about the rationale:

in the original test case we are dealing with three timeouts:

(1) Time we spend in thread_entry_fn:
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L211-L212

(2) Time we wait till the threads spawn on other cores:
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L336

(3) Time we delay creation of new thread (this one is even different for each thread)
https://github.com/zephyrproject-rtos/zephyr/blob/64894e8798d41bf2b6c8c3e36def394ba0a7bd82/tests/kernel/smp/src/main.c#L245-L253

On top of that there is some time we need to spawn new thread itself (4).

So we have multiple constraints which need to be true to have the test case logic correct:
max(3) < (1)
max(3) + (4) < (2)
min(3) + (4) + (1) > (2)
(2) < (1)
Note that 3 would differ for each thread and it maximum value depend on number of threads (number of cores)

The test case idea is to check that new coop thread with higher priority created won't preempt existing coop thread, and such dance with delays doesn't look robust at all.


> what's happened is that you've actually removed coverage of k_busy_wait()

That shouldn't decrease coverage of k_busy_wait:
 - we have tests for k_busy_wait (and even in `tests/kernel/smp` we still use k_busy_wait more than 13 times)
 - we don't call it with some specific parameter combinations
 - we don't call it in some specific context

So, I can remove it (or move it to separate PR), however I'm not really convinced here.

Reply:
This is making me less confident in your solution and not more.  Now I need to dig around and figure out exactly why that very large (250ms!!!!) delay is somehow not enough to spawn 12 threads...

Basically: are you 100% absolutely sure you aren't just working around a platform bug with k_busy_wait(), maybe because it's saturating a bus or timer device or something and preventing other CPUs from working?